uni-snd/src/native_snd.rs

```diff
@@ -0,0 +1,111 @@
+                        }
+                    }
+                    App::print(format!(
+                        "sound device : {} format {:?}\n",
+                        dev.name(),
+                        fmt.clone()
+                    ));
+                    format = Some(fmt);
+                    event_loop = Some(evt);
+                }
+                Err(e) => App::print(format!(
+                    "error : could not get default output format : {:?}\n",
+                    e
+                )),
+            }
+        } else {
+            App::print("warning : no sound device detected\n");
+        }
+        Self {
+            event_loop,
+            format,
+            stream_id,
+            tx: None,
+            generator: Some(generator),
+        }
+    }
+    pub fn send_event(&mut self, event: T) {
+        if let Some(ref mut tx) = self.tx {
+            tx.send(event).unwrap();
+        }
+    }
+    fn get_sample_rate(&self) -> f32 {
+        if let Some(ref fmt) = self.format {
+            fmt.sample_rate.0 as f32
+        } else {
+            1.0
+        }
+    }
+    pub fn frame(&mut self) {}
+    pub fn start(&mut self) {
+        let (tx, rx) = channel();
+        self.tx = Some(tx);
+        let stream_id = self.stream_id.take().unwrap();
+        let sample_rate = self.get_sample_rate();
+        let mut generator = self.generator.take().unwrap();
+        if let Some(evt) = self.event_loop.take() {
+            thread::spawn(move || {
+                App::print("starting audio loop\n");
+                evt.play_stream(stream_id);
+                generator.init(sample_rate);
+                evt.run(move |_stream_id, stream_data| {
+                    for event in rx.try_iter() {
+                        generator.handle_event(event);
+                    }
+
+                    match stream_data {
+                        cpal::StreamData::Output {
+                            buffer: cpal::UnknownTypeOutputBuffer::U16(mut buffer),
+                        } => for elem in buffer.iter_mut() {
+                            *elem = ((generator.next_value() * 0.5 + 0.5) * std::u16::MAX as f32)
+                                as u16;
+                        },
+                        cpal::StreamData::Output {
+                            buffer: cpal::UnknownTypeOutputBuffer::I16(mut buffer),
+                        } => for elem in buffer.iter_mut() {
+                            *elem = (generator.next_value() * std::i16::MAX as f32) as i16;
+                        },
+                        cpal::StreamData::Output {
+                            buffer: cpal::UnknownTypeOutputBuffer::F32(mut buffer),
+                        } => for elem in buffer.iter_mut() {
+                            *elem = generator.next_value();
+                        },
+                        _ => App::print(format!("unsupported stream data\n")),
```

Comment:
What should we do if it print out ? Could we `panic` or `unimplemented` here ?

Reply:
I dont think this can happen. If there exists an unsupported steam format, we should simply add code to support it, so yeah maybe we should rather panic here